File: src/gitsquid/clone.py
"""Getting a repository in the first place. Everything else in GitSquid needs one."""

from __future__ import annotations

import re
from pathlib import Path

from .gitcmd import GitError, both, run
from .refs import REMOTE_URL

CLONE_TIMEOUT = 900
NAME = re.compile(r"[A-Za-z0-9._-]{1,100}")


def repository_name(url: str) -> str:
    """`git@host:group/project.git` → `project`, which is where git would put it too."""
    tail = url.rstrip("/").rsplit("/", 1)[-1].rsplit(":", 1)[-1]
    return tail.removesuffix(".git") or "repository"


def clone(url: str, parent: Path, *, name: str = "") -> Path:
    """Clone into `parent/name`. The directory must not already exist."""
    url = (url or "").strip()
    if not REMOTE_URL.fullmatch(url) or len(url) > 2000:
        raise GitError("That does not look like a repository: use https://, ssh:// or user@host:path.")
    name = (name or "").strip() or repository_name(url)
    if not NAME.fullmatch(name):
        raise GitError("A folder name is letters, digits, dot, dash or underscore.")

    parent = Path(parent).expanduser()
    if not parent.is_dir():
        raise GitError(f"{parent} is not a directory.")
    target = parent / name
    if target.exists():
        raise GitError(f"{target} already exists. Choose another name, or open it.")

    result = run(parent, ["clone", "--", url, str(target)], timeout=CLONE_TIMEOUT)
    if result.returncode != 0:
        text = both(result)
        if "could not read Username" in text or "Authentication failed" in text or "Permission denied" in text:
            raise GitError(
                "Cloning needs credentials git could not supply without a prompt. "
                "Configure a credential helper or an SSH key, then try again."
            )
        raise GitError(f"Clone failed: {text or 'unknown error'}")
    return target

File: crates/gitsquid-core/src/clone.rs
//! Getting a repository in the first place. Everything else in GitSquid needs one.

use std::path::{Path, PathBuf};

use crate::git_cli::{run_in, CLONE};
use crate::refs::looks_like_a_remote_url;
use crate::{Error, Result};

/// `git@host:group/project.git` → `project`, which is where git would put it too.
pub fn repository_name(url: &str) -> String {
    let tail = url.trim_end_matches('/');
    let tail = tail.rsplit('/').next().unwrap_or(tail);
    let tail = tail.rsplit(':').next().unwrap_or(tail);
    match tail.strip_suffix(".git").unwrap_or(tail) {
        "" => "repository".to_string(),
        name => name.to_string(),
    }
}

fn is_folder_name(name: &str) -> bool {
    (1..=100).contains(&name.chars().count())
        && name.chars().all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '.' | '_' | '-'))
        && name != "."
        && name != ".."
}

/// Clone into `parent/name`. The directory must not already exist.
pub fn clone(url: &str, parent: &Path, name: &str) -> Result<PathBuf> {
    let url = url.trim();
    if !looks_like_a_remote_url(url) {
        return Err(Error::git(
            "That does not look like a repository: use https://, ssh:// or user@host:path.",
        ));
    }
    let name = match name.trim() {
        "" => repository_name(url),
        given => given.to_string(),
    };
    if !is_folder_name(&name) {
        return Err(Error::git("A folder name is letters, digits, dot, dash or underscore."));
    }
    if !parent.is_dir() {
        return Err(Error::git(format!("{} is not a directory.", parent.display())));
    }
    let target = parent.join(&name);
    if target.exists() {
        return Err(Error::git(format!(
            "{} already exists. Choose another name, or open it.",
            target.display()
        )));
    }

    let target_text = target.to_string_lossy().into_owned();
    let result = run_in(parent, &["clone", "--", url, &target_text], None, CLONE)?;
    if !result.ok() {
        let text = result.both();
        if text.contains("could not read Username")
            || text.contains("Authentication failed")
            || text.contains("Permission denied")
        {
            return Err(Error::git(
                "Cloning needs credentials git could not supply without a prompt. Configure a \
                 credential helper or an SSH key, then try again.",
            ));
        }
        return Err(Error::git(format!(
            "Clone failed: {}",
            if text.is_empty() { "unknown error" } else { &text }
        )));
    }
    Ok(target)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_name_is_taken_from_the_url_the_way_git_takes_it() {
        assert_eq!(repository_name("https://example.invalid/depot.git"), "depot");
        assert_eq!(repository_name("git@example.invalid:group/projet.git"), "projet");
        assert_eq!(repository_name("git@example.invalid:projet.git"), "projet");
        assert_eq!(repository_name("https://example.invalid/depot/"), "depot");
        assert_eq!(repository_name("https://example.invalid/depot"), "depot");
    }

    #[test]
    fn an_unnameable_url_falls_back() {
        assert_eq!(repository_name("/"), "repository");
        assert_eq!(repository_name(".git"), "repository");
    }

    #[test]
    fn folder_names_stay_inside_the_parent() {
        assert!(is_folder_name("depot"));
        assert!(is_folder_name("mon-depot_2.0"));
        for bad in ["", ".", "..", "a/b", "a b", "../escape", &"x".repeat(101)] {
            assert!(!is_folder_name(bad), "{bad:?} should be refused");
        }
    }
}
